src/main/java/io/github/fablabsmc/fablabs/api/fluid/v1/math/MixedNumber.java

```diff
@@ -0,0 +1,346 @@
+package io.github.fablabsmc.fablabs.api.fluid.v1.math;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.mojang.datafixers.Dynamic;
+import com.mojang.datafixers.types.DynamicOps;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.util.DynamicSerializable;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.google.common.math.IntMath.gcd;
+
+/**
+ * a number whose value is represented as 3 ints, a whole, numerator and denominator.
+ *
+ * <p>This object is immutable and self-simplifies on instantiation.
+ */
+public final class MixedNumber extends Number implements Comparable<MixedNumber>, DynamicSerializable {
+	public static final MixedNumber ZERO = new MixedNumber(0, 0, 1);
+	public static final MixedNumber ONE = new MixedNumber(1, 0, 1);
+
+	private final int whole;
+	private final int numerator;
+	private final /*Positive*/ int denominator;
+
+	/**
+	 * Should be only called if denominator is positive and numerator & denominator are coprime.
+	 */
+	private MixedNumber(int whole, int numerator, int denominator) {
+		if (denominator <= 0) throw new ArithmeticException("denominator cannot be less than or equal to 0");
+
+		if (gcd(Math.abs(numerator), denominator) != 1) {
+			throw new ArithmeticException("mixed number not simplified!");
+		}
+
+		this.whole = whole;
+		this.numerator = numerator;
+		this.denominator = denominator;
+	}
+
+	public static MixedNumber fromTag(CompoundTag tag) {
+		return of(tag.getInt("whole"), tag.getInt("numerator"), tag.getInt("denominator"));
+	}
+
+	public static MixedNumber of(int whole, int numerator, int denominator) {
+		if (denominator == 0) throw new ArithmeticException("Zero denominator");
+		return denominator < 0 ? ofValidDenominator(whole, -numerator, -denominator) : ofValidDenominator(whole, numerator, denominator);
+	}
+
+	//TODO: What shortcuts should we have, and how should we name them? This is in a Minecraft context, after all.
+	public static MixedNumber ofWhole(int numerator) {
+		if (numerator == 0) return ZERO;
+		if (numerator == 1) return ONE;
+		return new MixedNumber(0, numerator, 1);
+	}
+
+	public static MixedNumber ofThirds(int numerator) {
+		return ofValidDenominator(0, numerator, 3);
+	}
+
+	public static MixedNumber ofNinths(int numerator) {
+		return ofValidDenominator(0, numerator, 9);
+	}
+
+	public static MixedNumber ofThousandths(int numerator) {
+		return ofValidDenominator(0, numerator, 1000);
+	}
+
+	// should be only called if denom is positive
+	@VisibleForTesting
+	public static MixedNumber ofValidDenominator(int whole, int numerator, int denominator) {
+		if (numerator == 0) return new MixedNumber(whole, 0, 1);
+		if (numerator == denominator) return new MixedNumber(whole + 1, 0, 1);
```

Comment:
might be better to check `numerator > denominator`, and adjust the returned mixed number.
That way, for ex. `81/4` you'd only have a single step, rather than 20.

Reply:
ill do something about it